### PR TITLE
Fix new_heads Events Emission on Block Forks

### LINKED ISSUE #8848: newHeads ws subscription not publishing root header during reorg
#### System information

Erigon version: v2.54.0

OS & Version: Linux

Erigon Command (with flags/config):

```
--chain=mainnet
--http.addr=0.0.0.0
--http
--ws
--http.api=eth,debug,net,trace,web3,erigon
```

Consensus Layer Command (with flags/config):

```
--externalcl=false
```

Chain/Network: ethereum mainnet

#### Expected behaviour

if the reorg begins with 18673916, I expect 18673916 to be republished

18673915, 18673916, **18673916**, **18673917**, **18673918**

#### Actual behaviour

18673916 is not republished, and 18673917's parentHash points to a 18673916 header that was never received

18673915, 18673916, **18673917**, **18673918**

#### Steps to reproduce the behaviour

```golang
import (
  "github.com/ethereum/go-ethereum/ethclient"
  "github.com/ethereum/go-ethereum/rpc"
  "github.com/ethereum/go-ethereum/core/types"
)

rpcClient, err := rpc.Dial("ws://localhost:8545")
if err != nil {
  panic(err)
}
client := ethclient.NewClient(rpcClient)

var headerPC = make(chan *types.Header, 10)
client.SubscribeNewHead(context.Background(), headerPC)

for newHeader := range headerPC {
  fmt.Printf("new header received: %d\n\thash: %s\n\tparentHash: %s\n",
    newHeader.Number.Uint64(), newHeader.Hash().String(), newHeader.ParentHash.String(),
  )
}
```

#### golang log output

```
new header received: 18673915
  hash: 0xbe0b325962b2c29e8ead4ed70f0e62528856e8bf2659d49f8fc3c12b4b015cdd
  parentHash: 0x2fa1de31e4f1dafc3778aa60f0f97dd4cdf36279cce30421df7cf1160dca843e
new header received: 18673916
  hash: 0xc70ada05d79f36c9d17d762fc651c0d54fa7e62df417a8c6d806bd4179fab5c7
  parentHash: 0xbe0b325962b2c29e8ead4ed70f0e62528856e8bf2659d49f8fc3c12b4b015cdd
new header received: 18673917
  hash: 0xac48845a43c1158dc310b34c583328653028a69bd2c7d5007da233e73a6816f1
  parentHash: 0x8ea7a7ab7991b914b0013ace4129a91d37b135123e407d356929558ee778c80f
new header received: 18673918
  hash: 0x95ae4130df1613c1af9ea8d7e5949b5fcf94a4941a8862567256e8fe29caef43
  parentHash: 0xac48845a43c1158dc310b34c583328653028a69bd2c7d5007da233e73a6816f1
```

#### erigon logs

````
[INFO] [11-29|01:12:48.621] RPC Daemon notified of new headers       from=18673914 to=18673915 hash=0xbe0b325962b2c29e8ead4ed70f0e62528856e8bf2659d49f8fc3c12b...
[INFO] [11-29|01:12:48.621] head updated                             hash=0xbe0b325962b2c29e8ead4ed70f0e62528856e8bf2659d49f8fc3c12b4b015cdd number=18673915     
[INFO] [11-29|01:13:03.129] [NewPayload] Handling new payload        height=18673916 hash=0xc70ada05d79f36c9d17d762fc651c0d54fa7e62df417a8c6d806bd4179fab5c7     
[INFO] [11-29|01:13:03.638] [updateForkchoice] Fork choice update: flushing in-memory state (built by previous newPayload)                                       
[INFO] [11-29|01:13:04.058] RPC Daemon notified of new headers       from=18673915 to=18673916 hash=0xc70ada05d79f36c9d17d762fc651c0d54fa7e62df417a8c6d806bd41...
[INFO] [11-29|01:13:04.058] head updated                             hash=0xc70ada05d79f36c9d17d762fc651c0d54fa7e62df417a8c6d806bd4179fab5c7 number=18673916     
[INFO] [11-29|01:13:12.606] [NewPayload] Handling new payload        height=18673916 hash=0x8ea7a7ab7991b914b0013ace4129a91d37b135123e407d356929558ee778c80f     
[INFO] [11-29|01:13:13.101] [5/12 HashState] Unwinding started       from=18673916 to=18673915 storage=false codes=true                                          
[INFO] [11-29|01:13:13.102] [5/12 HashState] Unwinding started       from=18673916 to=18673915 storage=false codes=false                                         
[INFO] [11-29|01:13:13.105] [5/12 HashState] Unwinding started       from=18673916 to=18673915 storage=true codes=false                                          
[INFO] [11-29|01:13:13.111] [6/12 IntermediateHashes] Unwinding      from=18673916 to=18673915 csbucket=AccountChangeSet                                         
[INFO] [11-29|01:13:13.113] [6/12 IntermediateHashes] Unwinding      from=18673916 to=18673915 csbucket=StorageChangeSet                                         
[INFO] [11-29|01:13:13.146] [6/12 IntermediateHashes] Trie root      hash=0x9ae17ccebb28ab234900448c18571a746f1646bfda12c21c8307ba8b4212ec09                     
[INFO] [11-29|01:13:13.162] [4/12 Execution] Unwind Execution        from=18673916 to=18673915                                                                   
[INFO] [11-29|01:13:13.244] [4/12 Execution] Completed on            block=18673916                                                                              
[INFO] [11-29|01:13:13.517] head updated                             hash=0x8ea7a7ab7991b914b0013ace4129a91d37b135123e407d356929558ee778c80f number=18673916     
[INFO] [11-29|01:13:26.206] [NewPayload] Handling new payload        height=18673917 hash=0xac48845a43c1158dc310b34c583328653028a69bd2c7d5007da233e73a6816f1     
[INFO] [11-29|01:13:26.542] [updateForkchoice] Fork choice update: flushing in-memory state (built by previous newPayload)                                       
[INFO] [11-29|01:13:26.853] RPC Daemon notified of new headers       from=18673916 to=18673917 hash=0xac48845a43c1158dc310b34c583328653028a69bd2c7d5007da233e7...
[INFO] [11-29|01:13:26.853] head updated                             hash=0xac48845a43c1158dc310b34c583328653028a69bd2c7d5007da233e73a6816f1 number=18673917     
[INFO] [11-29|01:13:35.972] [NewPayload] Handling new payload        height=18673918 hash=0x95ae4130df1613c1af9ea8d7e5949b5fcf94a4941a8862567256e8fe29caef43  
````

### LINKED ISSUE #9568: Missing Block Headers During Chain Reorganization with SubscribeNewHeads on Erigon Versions 2.57.3 and 2.58
System information
Erigon version: at least 2.57.3 and 2.58

OS & Version: Linux

Expected behaviour
When subscribing to new block headers using the SubscribeNewHeads method (as described in the Geth documentation: RPC Pub/Sub newHeads), I expect to receive all new block headers in real-time, including those that are part of a chain reorganization.

Actual behaviour
The subscription works correctly for most new block headers. However, during a chain reorganization, the first block of the reorg is received, but subsequent blocks are not. This results in missing block headers for the blocks that are part of the reorganization. Here are the logs showing the received block headers, with a missing block in the sequence during a reorganization:

Command for erigon and its rpc:
* erigon --datadir=/mnt/bigdata/erigon-data --chain=mainnet --authrpc.addr 0.0.0.0 --authrpc.vhosts=* --http=false --torrent.download.rate=100mb --ws --private.api.addr=0.0.0.0:9090 --authrpc.jwtsecret=/mnt/bigdata/erigon-data/jwt.hex --nat any --log.console.verbosity info --metrics --metrics.addr 0.0.0.0 --maxpeers 200

* rpcdaemon --datadir=/mnt/bigdata/erigon-data --http.addr 0.0.0.0 --http.vhosts=* --http.api=eth,erigon,web3,net,debug,trace,txpool,engine,ots,admin --http.corsdomain=* --http.port=8545 --private.api.addr=0.0.0.0:9090 --log.console.verbosity info --ws --metrics --metrics.addr 0.0.0.0

Observe date-time between 19347342 (reorged) and 19347343 
2024/03/02 13:00:14 INFO New Block #19347341 -> hash: 0xb7d3984184200b203e9f25ffead478cb5a5c259f7ec3a8f9efdecf342cfb5829
2024/03/02 13:00:28 INFO New Block #19347342 -> hash: 0xefce71ab9437641251bfe076031e55f999b6f63b4d511cee05b7051b1b02dee3
2024/03/02 13:00:51 INFO New Block #19347343 -> hash: 0xc093b378ee0c6e5dc85deaa54a2f13f8d49ff6c5b19db78dada0718fcba9d50b
2024/03/02 13:01:01 INFO New Block #19347344 -> hash: 0x6edf3cc355133cedca5752dec53f0a520db29b26466dcbb9f2abcd315687ca07

Steps to reproduce the behaviour
1. Subscribe to new block headers using the SubscribeNewHeads method in a script or application interfacing with Erigon.
2. Wait for a chain reorganization event.
3. Observe the received block headers, noting any missing headers that should have been received as part of the reorganization.

### LINKED ISSUE #10056: Revisit new_heads Events Emission on Block Forks
See PR #9738

### PR DESCRIPTION
**Pull request by jotto**
TL;DR: on a reorg, the common ancestor block is not being published to subscribers of newHeads

#### Expected behavior

if the reorg's common ancestor is 2, I expect 2 to be republished

1, 2, **2**, **3**, **4**

#### Actual behavior

2 is not republished, and 3's parentHash points to a 2 header that was never received

1, 2, **3**, **4**

This PR is the same thing as https://github.com/ledgerwatch/erigon/pull/9738 except with a test.

Note... the test passes, but **this does not actually work in production** (for Ethereum mainnet with prysm as external CL).

Why? Because in production, `h.sync.PrevUnwindPoint()` is always nil: https://github.com/ledgerwatch/erigon/blob/a5270bccf5e69a6beaaab9a0663bdad80e989505/turbo/stages/stageloop.go#L291 which means the initial "if block" is never entered, and thus we have **no control** of increment/decrement `notifyFrom` during reorgs https://github.com/ledgerwatch/erigon/blob/a5270bccf5e69a6beaaab9a0663bdad80e989505/eth/stagedsync/stage_finish.go#L137-L146

I don't know why `h.sync.PrevUnwindPoint()` is seemingly always nil, or how the test can pass if it fails in prod. I'm hoping to pass the baton to someone who might. Thank you @indanielo for original fix.

If we can figure this bug out, it closes #8848 and closes #9568 and closes #10056

### CODE REVIEW

**Review by yperbasis — APPROVED** [review]

**Comment by Giulio2002** [comment]
Wait a second: https://geth.ethereum.org/docs/interacting-with-geth/rpc/pubsub

Here newHeads **only** republishes the last header of the chain. **NOT** the common ancestor and all of the subsequent blocks

**Comment by Giulio2002** [comment]
``
if the app requires handling reorgs, they need to find the common ancestor by themselves and reset their state
``

Directly from Geth team

**Comment by jotto** [comment]
Wow... color me surprised. Thank you for getting to the bottom of this @Giulio2002 

Someone with an ethereum.stackexchange account should update this post: https://ethereum.stackexchange.com/a/44930

image

**Comment by Giulio2002** [comment]
take a read, it is explained here: https://geth.ethereum.org/docs/interacting-with-geth/rpc/pubsub . newHeads only notify heads of the canonical chain. if something is in between, it is not supported
